modules/tensor_mechanics/src/materials/StrainEnergyDensity.C

```diff
@@ -0,0 +1,54 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "StrainEnergyDensity.h"
+#include "RankTwoTensor.h"
+#include "MooseMesh.h"
+
+template <>
+InputParameters
+validParams<StrainEnergyDensity>()
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription("Computes the strain energy density");
+  params.addParam<std::string>("base_name",
+                               "Optional parameter that allows the user to define "
+                               "multiple mechanics material systems on the same "
+                               "block, i.e. for multiple phases");
+  return params;
+}
+
+StrainEnergyDensity::StrainEnergyDensity(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Material>(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _strain_energy_density(declareProperty<Real>(_base_name + "strain_energy_density")),
+    _strain_energy_density_old(getMaterialPropertyOld<Real>(_base_name + "strain_energy_density")),
+    _stress(getMaterialProperty<RankTwoTensor>(_base_name + "stress")),
+    _stress_old(getMaterialPropertyOld<RankTwoTensor>(_base_name + "stress")),
+    _mechanical_strain(getMaterialPropertyByName<RankTwoTensor>(_base_name + "mechanical_strain")),
+    _mechanical_strain_old(
+        getMaterialPropertyOldByName<RankTwoTensor>(_base_name + "mechanical_strain"))
+{
+}
+
+void
+StrainEnergyDensity::initQpStatefulProperties()
+{
+  _strain_energy_density[_qp] = 0.0;
+}
+
+void
+StrainEnergyDensity::computeQpProperties()
+{
+  RankTwoTensor _my_strain_increment = (_mechanical_strain[_qp] - _mechanical_strain_old[_qp]);
```

Comment:
_my_strain_increment -> strain_increment (follow our convention that local variables don't have underscores).